### PR TITLE
Add `LevelMaintainerTerminal` rework logic inside `TileLevelMaintainer`

### PR DESCRIPTION
**Pull request by Laiff**
Close [GT-New-Horizons-Modpack#14554](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/14554)

Ready for review

1. Reworked registry to show in terminal
2. Reworked storage and scheduling inside LevelMaintaner (old storage format support migration on load)
3. Added new terminal to show level devices
4. Support back navigation inside terminals
5. Implemented InterfaceWireless using new solution 

Functionality of new terminal described in attached issue

### CODE REVIEW

**Comment by Laiff** [comment]
Final interface 
https://user-images.githubusercontent.com/575885/273729387-57486a36-21cd-4c90-a584-d9dfb1aee18f.png

**Comment by Dream-Master** [comment]
I just try to add your new pr but i run in issues and it wont build 
```
D:\Github\AE2FC\src\main\java\com\glodblock\github\client\gui\GuiFluidCraftConfirm.java:19: error: types com.glodblock.github.client.gui.IGuiTooltipHandler and appeng.client.gui.IGuiTooltipHandler are incompatible;
public class GuiFluidCraftConfirm extends GuiCraftConfirm implements IGuiTooltipHandler {
       ^
  class GuiFluidCraftConfirm inherits unrelated defaults for handleItemTooltip(ItemStack,int,int,List<String>) from types com.glodblock.github.client.gui.IGuiTooltipHandler and appeng.client.gui.IGuiTooltipHandler```
i just combined them with firenoos changes the other pr in the mod

File: src/main/java/com/glodblock/github/api/ISide.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.glodblock.github.api;
+
+import net.minecraftforge.common.util.ForgeDirection;
+
+public interface ISide {
+
+    ForgeDirection getSide();
```
Comment (miozune) [inline src/main/java/com/glodblock/github/api/ISide.java:7]: I guess AE2 is the right place to add this interface, like `AEBasePart` has a method with the same name

File: src/main/java/com/glodblock/github/network/wrapper/FCNetworkWrapper.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+ *  SimpleNetworkWrapper wrapper = NetworkRegistry.newSimpleChannel("MYCHANNEL");
+ *  // Message1 is handled by the Message1Handler class, it has discriminator id 1 and it's on the client
+ *  wrapper.registerMessage(Message1Handler.class, Message1.class, 1, Side.CLIENT);
+ *  // Message2 is handled by the Message2Handler class, it has discriminator id 2 and it's on the server
+ *  wrapper.registerMessage(Message2Handler.class, Message2.class, 2, Side.SERVER);
+ *  </pre>
+ * </code>
+ *
+ * @author cpw
+ */
```
Comment (miozune) [inline src/main/java/com/glodblock/github/network/wrapper/FCNetworkWrapper.java:87]: I think describing why this class is needed would be better than copying the entire Javadoc from `SimpleNetworkWrapper`

File: src/main/java/com/glodblock/github/common/tile/TileLevelMaintainer.java [hunk 3]
```diff
@@ -196,283 +214,406 @@ public void onChangeInventory(IInventory inv, int slot, InvOperation mc, ItemSta
...
 
-        public AeStackInventory<IAEItemStack> getRequestQtys() {
-            return this.requestQtys;
+        public void updateJob(int idx, Future<ICraftingJob> job) {
+            jobs[idx] = job;
         }
 
-        public AeStackInventory<IAEItemStack> getRequestBatches() {
-            return this.requestBatches;
+        private void updateField(int idx, Updater<NBTTagCompound, NBTTagCompound> updater) {
```
Comment (miozune) [inline src/main/java/com/glodblock/github/common/tile/TileLevelMaintainer.java:506]: `UnaryOperator` works

**Review by miozune — COMMENTED** [review]
Any reason for using different texture from AE2 for "highlight tile" button?

File: src/main/java/com/glodblock/github/util/Updater.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.glodblock.github.util;
+
+@FunctionalInterface
+public interface Updater<Input, Output> {
```
[the remark this answers is not in the document]
Comment (Laiff) [reply src/main/java/com/glodblock/github/util/Updater.java:4]: Main reason here is to have readable name for implementations and do not have `Function<>` somewhere in callstack. I can just inherit it with readable name from Function.

[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/com/glodblock/github/util/Updater.java:4]: I don't think it makes difference? Looking at the name "updater" doesn't tell us anything other than it updates something, which would be almost identical theme as `Function`.
`NBTUpdater` that is dedicated for taking and outputting NBT would be good though, as it clearly mentions what it does.

File: src/main/java/com/glodblock/github/api/ISide.java (hunk 1 above)
Comment (Laiff) [reply src/main/java/com/glodblock/github/api/ISide.java:7]: I plan to do some other thing related to terminals. I think I can move this interface after

File: src/main/java/com/glodblock/github/coremod/registries/LevelTerminalRegistry.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        this.register(PartLevelEmitter.class, new PartLevelEmitterAdapter());
+    }
+
+    public Set<Class<? extends ILevelViewable>> getSupportedClasses() {
+        return supportedClasses;
+    }
+
+    @Override
+    public boolean isAdopted(Class<? extends ILevelViewable> clazz) {
+        return adapterMap.containsKey(clazz);
```
[the remark this answers is not in the document]
Comment (Laiff) [reply src/main/java/com/glodblock/github/coremod/registries/LevelTerminalRegistry.java:42]: This is necessary in some cases it should use adapter functionality but in other cases it should use underlying machine itself.

**Comment by Laiff** [comment]
> Any reason for using different texture from AE2 for "highlight tile" button?

This icon more readable in small format and have better correlation with it function. Plan to replace it in AE2 too but not now

File: src/main/java/com/glodblock/github/util/Updater.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Laiff) [reply src/main/java/com/glodblock/github/util/Updater.java:4]: Interface completely removed

File: src/main/java/com/glodblock/github/coremod/registries/LevelTerminalRegistry.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/com/glodblock/github/coremod/registries/LevelTerminalRegistry.java:42]: Understood

File: src/main/java/com/glodblock/github/api/ISide.java (hunk 1 above)
Comment (miozune) [reply src/main/java/com/glodblock/github/api/ISide.java:7]: You mean you're going to move this interface later but not in this PR?

Comment (Laiff) [reply src/main/java/com/glodblock/github/api/ISide.java:7]: yes exactly

File: dependencies.gradle [hunk 6]
```diff
@@ -1,10 +1,44 @@
...
+ *
+ * For more details, see https://docs.gradle.org/8.0.1/userguide/java_library_plugin.html#sec:java_library_configurations_graph
+ */
 dependencies {
     api('com.github.GTNewHorizons:NotEnoughItems:2.4.5-GTNH:dev')
-    api('com.github.GTNewHorizons:Applied-Energistics-2-Unofficial:rv3-beta-262-GTNH:dev')
+    api('com.github.GTNewHorizons:Applied-Energistics-2-Unofficial:rv3-beta-271-GTNH-pre:dev')
     api('curse.maven:cofh-core-69162:2388751')
     api('com.github.GTNewHorizons:waila:1.6.2:dev')
+    api("com.github.GTNewHorizons:WirelessCraftingTerminal:1.10.1:dev")
```
Comment (miozune) [inline dependencies.gradle:41]: Duplicated entry with `compileOnly`. Also I think `implementation` is appropriate here, as other mods don't need reference to it while compile time. https://gtnh.miraheze.org/wiki/Development#Dependency_types

**Review by miozune — APPROVED** [review]
I'm not capable of going through every single change, but it looks good overall.
Update AE2 dep to non-pre one before merge.

**Pull request merged**